source/common/upstream/upstream_impl.cc

```diff
@@ -185,6 +186,8 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const std::vector<HostPtr>& n
   for (HostPtr host : new_hosts) {
     bool found = false;
     for (auto i = current_hosts.begin(); i != current_hosts.end();) {
+      zones.insert((*i)->zone());
```

Comment:
I think this line should be in the if statement below around line 199. Basically only insert when you push into final_hosts?